etc/compile_check.sh

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env bash
+set -e # exit when any command fails
+
+GC=go
+WD=$(pwd)
+COMPILE_CHECK_DIR="internal/test/compilecheck"
+DEV_MIN_VERSION=1.19
+
+# version will flatten a version string of upto 4 components for inequality
+# comparison.
+function version {
+	echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }';
+}
+
+# dev_compile_check will attempt to build in the development environment. This
+# check will only run on environments where the Go version is greater than or
+# equal to the DEV_MIN_VERSION.
+function dev_compile_check {
+	MACHINE_VERSION=`${GC} version | { read _ _ v _; echo ${v#go}; }`
+
+	if [ $(version $MACHINE_VERSION) -lt $(version $DEV_MIN_VERSION) ]; then
+		return
+	fi
+
+	GO111MODULE=on ${GC} build ./...
+	GO111MODULE=on ${GC} build ${BUILD_TAGS} ./...
+}
+
+# compile_check will attemps to build the the internal/test/compilecheck project
+# using the provided go version. This is to simulate a end-to-end use case. This
+# check will only run on environments where the Go version is greater than or
+# equal to the given version.
+function compile_check {
+	GO111MODULE=$1
+	VERSION=$2
+	MACHINE_VERSION=`${GC} version | { read _ _ v _; echo ${v#go}; }`
+
+	if [ $(version $MACHINE_VERSION) -lt $(version $VERSION) ]; then
+		return
+	fi
+
+	# Change the directory to the compilecheck test directory.
+	cd ${COMPILE_CHECK_DIR}
+
+	# If the version is not 1.13, then run "go mod tidy"
+	if [ "$VERSION" != 1.13 ]; then
+		go mod tidy -go=$VERSION
+	fi
+
+	# Check simple build.
+	${GC} build ./...
+
+	# Check build with dynamic linking.
+	${GC} build -buildmode=plugin
+
+	# Check build with tags.
+	go build $(BUILD_TAGS) ./...
+
+	# Check build with various architectures.
+	GOOS=linux GOARCH=386 ${GC} build $(BUILD_TAGS) ./...
+	GOOS=linux GOARCH=arm ${GC} build $(BUILD_TAGS) ./...
+	GOOS=linux GOARCH=arm64 ${GC} build $(BUILD_TAGS) ./...
+	GOOS=linux GOARCH=ppc64le ${GC} build $(BUILD_TAGS) ./...
+	GOOS=linux GOARCH=s390x ${GC} build $(BUILD_TAGS) ./...
```

Comment:
The `$(BUILD_TAGS)` in bash attempts to run the command `BUILD_TAGS` instead of doing a replacement like in Makefiles, leading to errors like:
> etc/compile_check.sh: line 57: BUILD_TAGS: command not found

Cross-compiling with build tags that enable C builds won't work on Evergreen CI hosts (or on local workstations usually), so it's best to just remove those.
```suggestion
	GOOS=linux GOARCH=386 ${GC} build ./...
	GOOS=linux GOARCH=arm ${GC} build ./...
	GOOS=linux GOARCH=arm64 ${GC} build ./...
	GOOS=linux GOARCH=ppc64le ${GC} build ./...
	GOOS=linux GOARCH=s390x ${GC} build ./...
```